pyteal/ast/router.py

```diff
@@ -0,0 +1,471 @@
+        - if it is handling abi-method, then
+          `Txn.application_arg[0] == hash(method_signature) &&
+           Txn.application_arg_num == 1 + min(METHOD_ARG_NUM_LIMIT, method's arg num)`
+          where `METHOD_ARG_NUM_LIMIT == 15`.
+        - if it is handling conditions for other cases, then `Txn.application_arg_num == 0`
+
+        Args:
+            method_signature: a string representing method signature for ABI method
+            method_to_register: an ABIReturnSubroutine if exists, or None
+            on_completes: a list of OnCompletion args
+            creation: a boolean variable indicating if this condition is triggered on creation
+        Returns:
+            approval_conds: A list of exprs for approval program's condition on: creation?, method/bare, Or[OCs]
+            clear_state_conds: A list of exprs for clear-state program's condition on: method/bare
+        """
+
+        # check that the onComplete has no duplicates
+        if len(on_completes) != len(set(on_completes)):
+            raise TealInputError(f"input {on_completes} has duplicated on_complete(s)")
+
+        # Check the existence of OC.CloseOut
+        close_out_exist = any(
+            map(lambda x: str(x) == str(OnComplete.CloseOut), on_completes)
+        )
+        # Check the existence of OC.ClearState (needed later)
+        clear_state_exist = any(
+            map(lambda x: str(x) == str(OnComplete.ClearState), on_completes)
+        )
+        # Ill formed report if app create with existence of OC.CloseOut or OC.ClearState
+        if creation and (close_out_exist or clear_state_exist):
+            raise TealInputError(
+                "OnComplete ClearState/CloseOut may be ill-formed with app creation"
+            )
+        # check if there is ABI method but no method_signature is provided
+        # TODO API change to allow inferring method_signature from method_to_register?
+        if method_to_register is not None and not method_signature:
+            raise TealInputError(
+                "A method_signature must be provided if method_to_register is not None"
+            )
+
+        # Check:
+        # - if current condition is for *ABI METHOD*
+        #   (method selector && numAppArg == 1 + min(METHOD_APP_ARG_NUM_LIMIT, subroutineSyntaxArgNum))
+        # - or *BARE APP CALL* (numAppArg == 0)
+        method_or_bare_condition = (
+            And(
+                Txn.application_args[0] == MethodSignature(cast(str, method_signature)),
+                Txn.application_args.length()
+                == Int(
+                    1
+                    + min(
+                        method_to_register.subroutine.argument_count(),
+                        METHOD_ARG_NUM_LIMIT,
+                    )
+                ),
+            )
+            if method_to_register is not None
+            else Txn.application_args.length() == Int(0)
```

Comment:
This would prevent any non ABI call from allowing arg passing?

Reply:
good point, I need to change this line since it is too strict, maybe the bare call with no subroutine-arg takes txn application args.

Visual reminder: I will circle myself back to this line after some API change.